Studio-Lovelies/GG-JointJustice-Unity
# Refactor Director Action Decoder to be easier to extend

## Summary
This is a PROPOSAL. It might get merged if people like the changes, and I will adjust as I get feedback, but if we decide as a team we don't want this change, I'll close the PR.

### Design Goals
- Make the decoding step testable
- Adding new commands is easy
    - No more writing your own parsing code for each command
- Automated, Correct, and Clear error logging if you get a command wrong
    - If you use an enum type, the error should output all possible enum values
    - If you enter 0 when the argument wanted a one-based integer, that should be an error
- Command parameters are easily discoverable
- (Experimental!) Docs write themselves! `ActionDecoder.WriteDocsForAllCommands()` returns a markdown doc with all the commands and the args they take.

### Changes
The switch case cases in DirectorActionDecoder (now moved to ActionDecoder) looks like the following:
`ActionName.PLAY_EMOTION: PlayEmotion(actionLine.NextString("emotion name"), actionLine.NextOptionalString("target actor"));`

This means `PlayEmotion` has a very clean implementation:
```
private void PlayEmotion(string poseName, string optional_targetActor)
{
    if (!HasActorController())
        return;

    if (optional_targetActor == null)
    {
        ActorController.PlayEmotion(poseName);
    }
    else
    {
        ActorController.PlayEmotion(poseName, optional_targetActor);
    }
}
```

### Notable Findings
- SetActorPosition used to log an error explaining it's usage that's actually incorrect (the args were in the wrong order). This would have been really hard to notice and is exactly the type of problem I'm trying to fix.
- To test a command in isolation (eg: `"&SET_ACTOR_POSITION:3,Arin"` you actually need to end with a space: `"&SET_ACTOR_POSITION:3,Arin "`). This is probably a bug but I didn't fix it because this is a Refactor PR.

### Drawbacks to this approach
Although using the API (aka: creating a new command) is much easier, the underlying logic is significantly more complicated. This won't matter to most people on the team, but if you want an `OptionaBool` for instance you'll have to figure out how to add it (it's not hard to figure out, but it's learning you didn't need to do before).

If you provide _too many_ arguments there is no longer a mechanism to detect that. `&SET_ACTOR_POSTIION:3,Arin,butts,360noscope` will not error. The extra args will be ignored.

## Testing instructions
- Test all the features you know about to make sure I didn't break anything (93% confident it's safe)
- Try to enter commands that don't exist or have improper arguments and see if you get a helpful error message

## Additional information

- `[ ]` Changes UI
- `[ ]` Introduces new feature
- `[ ]` Removes existing feature
- `[ ]` Has associated resource:
